README.md
# Xhh_Auto

[![Codacy Badge][codacy_b]][Codacy] [![release][release_b]][Release] [![PyPI][pypi_v_b]][pypi] [![Download][download_b]][Release] [![License][license_b]][License]

## 声明

    本项目仅供学习交流使用, 用户如何使用该脚本与作者无关

* 请从这里下载稳定版: [Releases][release], 仓库中的代码未经测试, 可能会有问题.

## Frida_RPC服务部署方法

![demo](https://blog.chrxw.com/usr/uploads/2021/11/59198763.png)

* Frida RPC简介：[这篇文章](https://blog.chrxw.com/archives/2021/11/19/1640.html)

* RPC Server获取：[frida_project](https://github.com/chr233/frida_project)

## 安装与使用方法

    本项目使用`Python3.8`编写,理论上支持`3.4`及以后的版本

1. 从[releases][release]下载最新版本.
2. 解压,并切换到脚本所在目录.
3. 复制 `example.config.toml` , 为 `config.toml`,并填写配置.
    > 账号凭据获取方法参考下一节.
4. 在脚本目录打开命令行.
5. 执行 `pip install -r ./requirements.txt`.
6. 运行脚本: windows `python run.py`, linux: `python3 run.py`
7. 本脚本没有定时运行的功能,如有需要,请参考下文

## 账号凭据获取方法

    当前版本暂不支持直接登录,需要手动抓包获取账号凭据,抓取后只要不主动退出可以永久登录

* [安卓手机端抓取教程【新】][jiaocheng1]【推荐】

* [安卓手机抓取教程][jiaocheng2]

* [Fiddler抓取教程][jiaocheng3]

* 抓取小黑盒数据后, 填入`config.toml`即可使用本脚本.

## 旧版脚本配置文件转换

* 运行`python convernt_old_config.py`,根据提示操作即可.

## 定时运行方法

    因为本脚本没有必要常驻后台,所以没有自动运行的功能,不过可以配合系统工具达到自动运行的效果

* Windows下可以使用`任务计划程序`
    1. 开始菜单搜索`任务计划程序`
    2. 点右侧`创建基本任务…`
    3. `名称`和`描述`任意填
    4. `触发器`选择`每天`,并设置运行时间
    5. `操作`选择`启动程序`,`程序或脚本`填`python [脚本路径]\run.py`
    6. 点`完成`

* Linux下可以使用`crontab`
    1. 打开终端,输入`corntab -e`
    2. 按`i`进入编辑模式,将光标定位到文件尾
    3. 插入`0 1 * * * python3 [脚本绝对路径]/run.py > /dev/null 2>&1 &`
        > 这个配置的意思是每天1点0分执行脚本
    4. 按`ESC`,然后输入`:wq`保存

## XHH_AUTO 互助计划

    参与互助计划以后,会自动关注我的账号以及关注我的粉丝的账号,也就是让使用本脚本的用户都能互相关注,再配合动态点赞功能,让所有用户都能快速获赞,涨粉,升级

> 退出方法:在`config.toml`中将`join_xhhauto`的值改为`false`即可.

[codacy_b]: https://app.codacy.com/project/badge/grade/dfb3196838bf4431a8914736f103afeb
[codacy]: https://www.codacy.com/manual/chr233/xhh_auto?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=chr233/xhh_auto&amp;utm_campaign=badge_grade
[download_b]: https://img.shields.io/github/downloads/chr233/xhh_auto/total
[release]: https://github.com/chr233/xhh_auto/releases
[release_b]: https://img.shields.io/github/v/release/chr233/xhh_auto
[license]: https://github.com/chr233/xhh_auto/blob/master/license
[license_b]: https://img.shields.io/github/license/chr233/xhh_auto
[pypi_v_b]: https://img.shields.io/pypi/v/pyxiaoheihe?label=pyxiaoheihe
[pypi]: https://pypi.org/project/pyxiaoheihe/
[jiaocheng1]: https://blog.chrxw.com/archives/2020/08/10/1353.html
[jiaocheng2]: https://blog.chrxw.com/archives/2019/10/19/390.html
[jiaocheng3]: https://blog.chrxw.com/archives/2019/10/20/437.html


CHANGELOG.md
# 更新日志

## 2021-11-19 v0.93

* 改为使用Frida RPC计算hkey, 需要在config.toml中heybox节内定义rpc_server, 并部署RPC server
* 模拟小黑盒版本为1.3.191
* `pyxiaoheihe`版本需要>=1.0.11

## 2020-09-04 v0.91

* 支持为每个账号配置不同的请求头, 方便多设备用户使用
* 添加了对`点赞次数用尽`这个错误的处理
* 稍微修改了下消息汇报格式
* `pyxiaoheihe`版本需要>=1.0.8

## 2020-08-28 v0.90

* `pyxiaoheihe`作为模块从本项目中独立出去, 可以使用`pip install pyxiaoheihe --upgrade`安装和更新.
* 修复了几个小问题.

## 2020-08-21 v0.90

* 配置文件有更新, 详细内容请查看`example.config.toml`

* 适配小黑盒`1.3.121`客户端的加密算法
* 增加了数据上报模块, 模拟正常客户端行为
* 增加了随机延时机制(可以在`config.toml`中修改延时倍率)
* 修正数个小问题

## 2020-08-07 v0.85

* 修复上个版本的几处错误
* 添加了发送评论的方法
* 隐藏用法请加群

## 2020-08-05 v0.84

### 注意事项

* 该版本新增了2个依赖库, 请使用命令 `pip isntall -r requirements.txt` 重新安装依赖
* 该版本修改了配置文件, 建议对比 `example.config.toml` 手动更新

### 更新内容

* 全面支持小黑盒加密协议, 隐藏用法请加群
* 从该版本开始发行版中不再附带 `convernt_old_config.py`
* 新支持了几条API

## 2020-08-04 v0.83

* 修复了拉取动态不能正确判断有没有点过赞的问题
* 更新提醒现在会附在执行结果后面而不是单独推送

## 2020-08-01 v0.82

* 重构了原来的代码, 更加简洁高效
* 使用全新的配置文件格式, 可以使用 `convernt_old_config.py` 进行转换
* 当前 `pyxiaoheihe` 只实现了完成每日任务所必须的API, 待完成度比较高以后将成为独立的模块
* `xhh_auto` 这个项目将专注于自动化运行, `pyxiaoheihe` 将专注于完整实现手机客户端的功能
* 推出了 `xhhauto互助计划` , 具体信息请参考 `README.md`
* 有问题欢迎提交issues, 或者在[我的博客](https://blog.chrxw.com)上留言

## 2020-07-29 v0.79

* 更新新版hkey机制
* 大部分沿用旧版代码, 后续将进行重构

## 【以下版本均已失效, 仅作存档】

## 2019-11-23 v0.70

* 新增 `get_user_comment_list` 函数, 可以拉取用户评论, **暂不能判断是否已赞**
* 新增 `batch_like_commentlist` 函数, 可以批量点赞评论列表
* 默认不解析HTML, 手机端执行不会因为缺少lxml报错
* 修改了几个批量拉取函数, 执行速度提升约20%
* 修改了 `start.py` 中的几处语法错误
* 小幅调整控制台的了输出样式
* ~~下一个大版本将会重构~~

## 2019-10-18 v0.60

* 修改了一处输出错误
* 新增 `get_news_comments` 函数, 可以批量拉取文章评论
* 新增 `share_comment` 函数, 用于分享文章评论
* 函数修改 `get_daily_task_detail` , 现在能正确检测4项每日任务是否完成
* 函数更名 `share` -> `share_news` , 原函数将被删除
* 函数更名 `get_follower_list` -> `get_follower_list_by_userid` , 原函数将被删除, 新函数支持获取他人的粉丝列表
* 函数更名 `get_following_list` -> `get_following_list_by_userid` , 原函数将被删除, 新函数支持获取他人的关注列表
* 现在可以完成小黑盒新的每日任务了(分享评论), 每日经验+10

## 2019-10-04 v0.50

* 增加了更新检测的开关( `UpdateCheck` ), 不过不建议关闭
* 修复了直接运行脚本时读取配置文件的路径不正确的问题
* 增加了运行结束后的提示, 使用参数-N禁用
* 修复了几个bug

## 2019-9-29 v0.48

* 修复了检查脚本更新会闪退的bug
* 加了一层错误捕获机制
* 增加了方糖推送的开关(EnableFtqq)

## 2019-9-26 v0.45

* 修复了几个错误(然而还是没修完)
* 小幅修改了输出格式

## 2019-9-25 v0.4

* 重构了全部代码, 根据功能拆分模块
* 文档注释规范化, 为所有函数添加了统一的注释
* 完善了错误处理, 精简了log输出
* 可以获取指定用户的动态、发帖、评论
* 修改部分API入口地址
* 重命名部分常量, 使描述更加清晰
* 精简了一些只能单次操作的函数, 用能批量操作的函数替代
* 更加友好的输出格式

## 2019-8-9 v0.3

* 可以自动关注推荐关注列表中的用户, 并且可以过滤关注数和粉丝数差距过大的用户
* 可以清理单向关注的用户(只清理关注数和粉丝数差距过大的用户)
* 可以对好友发送私信(只能发)
* ~~添加了N个没什么存在感被我忘记的的函数~~

## 2019-7-30 v0.2

* 添加了对方糖(Server酱)的支持, 可以将脚本结果推送给微信
* 可以自动关注自己的粉丝
* 逆向出了hkey的计算方法~~(之前是瞎算的)~~, 正确的hkey应该能减少被检测到的概率

## 2019-7-22 v0.1

* 简单支持多账号
* 可以自动完成每日任务(签到, 点赞, 分享)
* 可以自动给好友动态点赞
* 可以获取账户详细数据, 拉取小黑盒新闻


utils/config.py

'''
# @Author       : Chr_
# @Date         : 2020-07-29 14:21:39
# @LastEditors  : Chr_
# @LastEditTime : 2021-11-19 11:56:29
# @Description  : 读取并验证配置
'''

import os
import toml
import chardet
from .log import get_logger, init_logger

logger = get_logger('Setting')

SCRIPT_PATH = f'{os.path.split(os.path.realpath(__file__))[0][:-5]}'

DEFAULT_PATH = f'{SCRIPT_PATH}config.toml'

CFG = {}


def get_script_path() -> str:
    '''
    获取脚本所在路径

    返回:
        str: 脚本所在路径
    '''
    return(SCRIPT_PATH)


def get_config(key: str) -> dict:
    '''
    获取某一项配置

    参数:
        key: 要获取的设置键名
    返回:
        dict: 配置信息字典
    '''
    return(CFG.get(key))


def get_all_config() -> dict:
    '''
    获取全部配置

    返回:
        dict: 配置信息字典
    '''
    return(CFG)


def load_config(path: str = DEFAULT_PATH) -> dict:
    '''
    读取并验证配置

    参数:
        [path]: 配置文件路径,默认为config.toml
    返回:
        dict: 验证过的配置字典
    '''
    global CFG
    try:
        logger.debug('开始读取配置')
        with open(path, 'rb') as f:
            content = f.read()
        detect = chardet.detect(content)
        encode = detect.get('encode', 'utf-8')
        raw_cfg = dict(toml.loads(content.decode(encode)))
        CFG = verify_config(raw_cfg)
        debug = os.environ.get('mode', 'release').lower()
        level = 0 if debug == 'debug' else 20
        init_logger(level)
        logger.debug('配置验证通过')
        return(CFG)

    except FileNotFoundError:
        logger.error(f'[*] 配置文件[{path}]不存在')
        raise FileNotFoundError(f'[*] 配置文件[{path}]不存在')
    except ValueError as e:
        logger.error(f'[*] 配置文件验证失败 [{e}]', exc_info=True)


def verify_config(cfg: dict) -> dict:
    '''
    验证配置

    参数:
        cfg: 配置字典
    返回:
        dict: 验证过的配置字典,剔除错误的和不必要的项目
    '''
    vcfg = {'main': {'check_update': True, 'debug': False,  'join_xhhauto': True},
            'ftqq': {'enable': False, 'skey': '', 'only_on_error': False},
            'email': {'port': 465, 'server': '', 'password': '', 'user': '',
                      'recvaddr': '', 'sendaddr': '', 'only_on_error': False},
            'heybox': {'channel': 'heybox_yingyongbao', 'os_type': 'Android',
                       'os_version': '9', 'sleep_interval': 1.0, 'auto_report': True},
            'accounts': []}

    main = cfg.get('main', {})
    if main and isinstance(main, dict):
        check_update = bool(main.get('check_update', True))
        debug = bool(main.get('debug', False))
        join_xhhauto = bool(main.get('join_xhhauto', True))
        vcfg['main'] = {'check_update': check_update, 'debug': debug,
                        'join_xhhauto': join_xhhauto}
    else:
        logger.debug('[main]节配置有误或者未配置,将使用默认配置')

    ftqq = cfg.get('ftqq', {})
    if ftqq and isinstance(ftqq, dict):
        enable = bool(ftqq.get('enable', False))
        skey = ftqq.get('skey', "")
        only_on_error = bool(ftqq.get('only_on_error', False))
        if enable and not skey:
            raise ValueError('开启了FTQQ模块,但是未指定SKEY,请检查配置文件')
        vcfg['ftqq'] = {'enable': enable, 'skey': skey,
                        'only_on_error': only_on_error}
    else:
        logger.debug('[ftqq]节配置有误或者未配置,将使用默认配置')

    email = cfg.get('email', {})
    if email and isinstance(email, dict):
        enable = bool(email.get('enable', False))
        try:
            port = int(email.get('port', 0))
        except ValueError:
            port = 465
            logger.warning('[*] [email]节port必须为数字')
        server = email.get('server', '')
        password = email.get('password', '')
        user = email.get('user', '')
        recvaddr = email.get('recvaddr', '')
        sendaddr = email.get('sendaddr', '')
        only_on_error = bool(email.get('only_on_error', False))
        if enable and not (port and server
                           and password and user and recvaddr and sendaddr):
            raise ValueError('开启了email模块,但是配置不完整,请检查配置文件')
        vcfg['email'] = {'enable': enable, 'port': port, 'server': server,
                         'password': password, 'user': user,
                         'recvaddr': recvaddr, 'sendaddr': sendaddr,
                         'only_on_error': only_on_error}
    else:
        logger.debug('[email]节配置有误或者未配置,将使用默认配置')

    heybox = cfg.get('heybox', {})
    if heybox and isinstance(heybox, dict):
        channel = heybox.get('channel', "heybox_yingyongbao")
        try:
            os_type = int(heybox.get('os_type', 1))
            if os_type not in (1, 2):
                raise ValueError
        except ValueError:
            os_type = 1
            logger.warning('[*] [heybox]节os_type只能为1或者2')
        os_version = heybox.get('os_version', "9")
        try:
            sleep = float(heybox.get('sleep_interval', 1))
        except ValueError:
            sleep = 1.0
            logger.warning('[*] [heybox]节sleep_interval必须为数字')
        auto_report = bool(heybox.get('auto_report', False))
        rpc_server = heybox.get('rpc_server', 'http://localhost:9000/encode')

        vcfg['heybox'] = {'channel': channel, 'os_type': os_type,
                          'os_version': os_version, 'sleep_interval': sleep,
                          'auto_report': auto_report, 'rpc_server': rpc_server}
    else:
        logger.debug('[heybox]节配置有误或者未配置,将使用默认配置')

    accounts = cfg.get('accounts', [])
    if accounts and isinstance(accounts, list):
        vcfg['accounts'] = []
        for i, account in enumerate(accounts, 1):
            try:
                heybox_id = int(account['heybox_id'])
                imei = account['imei']
                pkey = account['pkey']
                i_os_type = account.get('os_type') or os_type
                i_channel = account.get('channel') or channel
                i_os_version = account.get('os_version') or os_version

                if heybox_id and imei and pkey:
                    vcfg['accounts'].append({'heybox_id': heybox_id, 'imei': imei,
                                             'pkey': pkey, 'os_type': i_os_type,
                                             'channel': i_channel, 'os_version': i_os_version})
                else:
                    raise ValueError
            except (ValueError, AttributeError):
                logger.warning(f'[*] 第{i}项账号配置有误,已忽略该项')
                logger.debug(f'[*] 配置项为{account}')

    if not vcfg['accounts']:
        logger.error('[*] 不存在有效的账号信息,请检查config.toml')
    return(vcfg)
